pym/bob/archive.py

```diff
@@ -0,0 +1,389 @@
+        except urllib.error.URLError as e:
+            print(colorize(str(e.reason), "33"))
+        except OSError as e:
+            raise BuildError("Error: " + str(e))
+        finally:
+            urllib.request.urlcleanup()
+
+        return ret
+
+    def upload(self, step, buildIdFile, tgzFile):
+        # only upload if requested
+        if not self.canUpload():
+            return ""
+
+        # only upload tools if built in sandbox
+        if step.doesProvideTools() and (step.getSandbox() is None):
+            return ""
+
+        # upload with curl if file does not exist yet on server
+        return "\n" + textwrap.dedent("""\
+            # upload artifact
+            cd $WORKSPACE
+            BOB_UPLOAD_URL="{URL}/$(hexdump -e '2/1 "%02x/" 14/1 "%02x"' {BUILDID}).tgz"
+            if ! curl --output /dev/null --silent --head --fail "$BOB_UPLOAD_URL" ; then
+                curl -sSg -T {RESULT} "$BOB_UPLOAD_URL" || echo Upload failed: $?
+            fi""".format(URL=self.__url, BUILDID=quote(buildIdFile), RESULT=quote(tgzFile)))
+
+    def download(self, step, buildIdFile, tgzFile):
+        # only download if requested
+        if not self.canDownload():
+            return ""
+
+        # only download tools if built in sandbox
+        if step.doesProvideTools() and (step.getSandbox() is None):
+            return ""
+
+        return "\n" + textwrap.dedent("""\
+            BOB_DOWNLOAD_URL="{URL}/$(hexdump -e '2/1 "%02x/" 14/1 "%02x"' {BUILDID}).tgz"
+            curl -sSg --fail -o {RESULT} "$BOB_DOWNLOAD_URL" || echo Download failed: $?
+            """.format(URL=self.__url, BUILDID=quote(buildIdFile), RESULT=quote(tgzFile)))
+
+
+class CustomArchive(BaseArchive):
+    """Custom command archive"""
+
+    def __init__(self, spec, whiteList):
+        super().__init__(spec)
+        self.__downloadCmd = spec.get("download")
+        self.__uploadCmd = spec.get("upload")
+        self.__whiteList = whiteList
+
+    def _makeUrl(self, buildId):
+        packageResultId = asHexStr(buildId)
+        return "/".join([packageResultId[0:2], packageResultId[2:4],
+            packageResultId[4:] + ".tgz"])
+
+    def canDownload(self):
+        return super().canDownload() and (self.__downloadCmd is not None)
+
+    def canUpload(self):
+        return super().canUpload() and (self.__uploadCmd is not None)
+
+    def uploadPackage(self, buildId, path):
+        if not self.canUpload():
+            return
+
+        print(colorize("   UPLOAD    {}".format(path), "32"))
+        (tmpFd, tmpName) = mkstemp()
+        try:
+            os.close(tmpFd)
+            with tarfile.open(tmpName, mode="w:gz") as tar:
+                tar.add(path, arcname=".")
+
+            env = { k:v for (k,v) in os.environ.items() if k in self.__whiteList }
+            env["BOB_LOCAL_ARTIFACT"] = tmpName
+            env["BOB_REMOTE_ARTIFACT"] = self._makeUrl(buildId)
+            ret = subprocess.run(["/bin/bash", "-ec", self.__uploadCmd],
```

Comment:
Maybe I'm wrong but this limits uploadCmd to a cmd given in default.yaml? I'd prefer to put my commands into a separate shell script and/or have the full string functions. E.g.:
```yaml
archive:
       -
           backend: shell
           upload: "$<<scripts/myscript>> upload ${BOB_LOCAL_ARTIFACT} ${BOB_REMOTE_ARTIFACT}"
           upload: "$<<scripts/myscript>> download ${BOB_LOCAL_ARTIFACT} ${BOB_REMOTE_ARTIFACT}"
```